In [191]:
import numpy as np
import pandas as pd


# Load the dataset

In [153]:
df = pd.read_excel('CIBC_App_Reviews.xlsx')
pd.options.display.max_columns = None

# Explore the dataset

In [164]:
# take a quick look at the data
df.head(10)

App Name App Store                  App Store     App ID  \
0  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
1  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
2  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
3  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
4  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
5  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
6  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
7  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
8  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
9  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   

    Review ID Country  Version  Rating       Date  \
0  3902085253  Canada  10.10.3       1 2019-03-20   
1  3901178898  Canada  10.10.3       2 2019-03-19   
2  3893450681  Canada  10.10.3       3 2019-03-17   
3  3882357860   China  10.10.3       1 2019-03-15   
4  3885434139  Canada  10.10.3       1 2019-03-15   
5  3881274434  Canada  10.10.3       1 2019-03-14   
6  3865788153  Canada  10.10.3       5 2019-03-10   
7  3866472114  Canada  10.10.3       1 2019-03-10   
8  3853019071  Canada  10.10.3       1 2019-03-07   
9  3847909796  Canada  10.10.3       1 2019-03-06   

                        Author                               Subject  \
0              rbc for the win                          Switch banks   
1              Leaving Wattpad              Credit Score Maintenance   
2               CanadianBatman                                  Good   
3                          孙三皮                                怎么切换语言   
4  allthenicknameseveraretaken                      Not good at all.   
5                      Binketh                               Log out   
6                 Leo 990@8888                      Cash back on app   
7               Jolena Katusha                    would give 0 stars   
8      Big Bird Below The Belt  Ve r    y    S  l   o      w    now😴   
9               Kiikikikikokik               theres always problems!   

                                                Body       Translated Subject  \
0        This app is as crappy as the bank that made                      NaN   
1  The credit score calculator has been down for ...                      NaN   
2  I stopped banking with CIBC due to another ban...                      NaN   
3                                             怎么是法语？  How to switch languages   
4  I have to change my password every time I try ...                      NaN   
5  Can’t log out. Or is very cumbersome to log ou...                      NaN   
6  I love the app. I never faced any issues so fa...                      NaN   
7  The absolute worst banking experience i’ve eve...                      NaN   
8  Very slow since last review, shield with the c...                      NaN   
9  it frequently wont allow me to transfer money ...                      NaN   

  Translated Body Sentiment  Device              Language  OS Version  \
0             NaN  negative     NaN               English         NaN   
1             NaN  negative     NaN               English         NaN   
2             NaN  positive     NaN               English         NaN   
3  How is French?  negative     NaN  Chinese (Simplified)         NaN   
4             NaN  negative     NaN               English         NaN   
5             NaN  negative     NaN               English         NaN   
6             NaN  positive     NaN               English         NaN   
7             NaN  negative     NaN               English         NaN   
8             NaN  negative     NaN               English         NaN   
9             NaN  negative     NaN               English         NaN   

                                           Reply URL  \
0  https://app.appbot.co/apps/30102-cibc-mobile-b...   
1  https://app

** Some Observations**
<br>
1) It seems both Subject and Body have texts, we would need to combine them for analysis.
<br>
2) We would need to remove emojis.
<br>
3) Not all reviews are in English so we could either translate them into English or remove them for now.
<br>
Continue our exploration...

** Quick Stats**
<br>
Get some quick stats, for example, count of ratings per rating score, rating averages, and ratings by language.

In [166]:
print('--Check total records and columns--\n')
shape = df.shape
print(shape)
print('\nThe dataset has {} rows and {} columns.\n'.format(shape[0],shape[1]))

print('--Check the total # of empty values for each variable--\n')

print(df.isnull().sum())

--Check total records and columns--

(7738, 23)

The dataset has 7738 rows and 23 columns.

--Check the total # of empty values for each variable--

App Name                 0
App Store                0
App                      0
Store                    0
App ID                   0
Review ID              390
Country                  0
Version               4646
Rating                   0
Date                     0
Author                 340
Subject               2219
Body                    29
Translated Subject    7681
Translated Body       7552
Sentiment                0
Device                7738
Language                 7
OS Version            7738
Reply URL                0
Topics                   0
Custom Topics            0
Tags                     0
dtype: int64


Subject and Body have 2,219 and 29 empty values respectively. We would need to combine them and see if we still see empty values for the combined field.
<br>


In [186]:
print('--Get the rating count per rating--\n')
rating_count = df.groupby(['Rating'])['Rating'].count()\
                                            .reset_index(name='Count')
rating_count['Percent'] = rating_count.Count/rating_count.Count.sum()
print(rating_count.sort_values('Percent',ascending=False))

print('\n--Get the rating count per language--\n')
lan_count = df.groupby(['Language']).agg({'Language' : 'count', 'Rating' : 'mean'})\
                                    .rename(columns={'Language':'Total_Reviews','Rating' : 'Rating_Avg'}).reset_index()
lan_count['Percent'] = lan_count.Total_Reviews/lan_count.Total_Reviews.sum()
print(lan_count.sort_values('Percent',ascending=False))



--Get the rating count per rating--

   Rating  Count   Percent
4       5   2581  0.333549
0       1   2442  0.315585
3       4   1046  0.135177
2       3    862  0.111398
1       2    807  0.104291

--Get the rating count per language--

                 Language  Total_Reviews  Rating_Avg   Percent
3                 English           7400    3.049865  0.957185
4                  French            293    3.399317  0.037899
1    Chinese (Simplified)             15    2.266667  0.001940
0                  Arabic              7    4.142857  0.000905
12                Spanish              4    4.000000  0.000517
7                  Korean              3    4.333333  0.000388
2   Chinese (Traditional)              2    4.000000  0.000259
5              Indonesian              2    5.000000  0.000259
6                 Italian              1    5.000000  0.000129
8                   Latin              1    4.000000  0.000129
9                   Malay              1    5.000000  0.000129
10   

1) 1/3 of reviews are either 5 or 1. Close to half of (46%) reviews are good (score 4 or 5) and approximately 40% of reviews are bad (score 1 or 2). It seems users who left comment either love the app or hate it.
<br>
2) Most of reviews (96%) are English reviews. 7400 is good enough for the analysis. I will discard reviews in other languages for now.
<br>
3) People who leave English reviews think the app is okay giving a rate of 3.05 on average.

# Data Cleansing

Here are the few steps before we can train model on this dataset.
<br>
1) Keep English reviews only.
<br>
2) Select the columns we need.
<br>
3) Remove punctuations, emojis.
<br>
4) Convert all the words into lowercase.

In [306]:
print('--1) Keep only English reviews--\n')
Eng_reviews = df[df['Language'] =='English']
print('Now the total of reviews becomes {}.\n'.format(len(Eng_reviews)))
Eng_reviews.head()

--1) Keep only English reviews--

Now the total of reviews becomes 7400.



App Name App Store                  App Store     App ID  \
0  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
1  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
2  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
4  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   
5  CIBC Mobile Banking       iOS  CIBC Mobile Banking   iOS  351448953   

    Review ID Country  Version  Rating       Date  \
0  3902085253  Canada  10.10.3       1 2019-03-20   
1  3901178898  Canada  10.10.3       2 2019-03-19   
2  3893450681  Canada  10.10.3       3 2019-03-17   
4  3885434139  Canada  10.10.3       1 2019-03-15   
5  3881274434  Canada  10.10.3       1 2019-03-14   

                        Author                   Subject  \
0              rbc for the win              Switch banks   
1              Leaving Wattpad  Credit Score Maintenance   
2               CanadianBatman                      Good   
4  allthenicknameseveraretaken          Not good at all.   
5                      Binketh                   Log out   

                                                Body Translated Subject  \
0        This app is as crappy as the bank that made                NaN   
1  The credit score calculator has been down for ...                NaN   
2  I stopped banking with CIBC due to another ban...                NaN   
4  I have to change my password every time I try ...                NaN   
5  Can’t log out. Or is very cumbersome to log ou...                NaN   

  Translated Body Sentiment  Device Language  OS Version  \
0             NaN  negative     NaN  English         NaN   
1             NaN  negative     NaN  English         NaN   
2             NaN  positive     NaN  English         NaN   
4             NaN  negative     NaN  English         NaN   
5             NaN  negative     NaN  English         NaN   

                                           Reply URL  \
0  https://app.appbot.co/apps/30102-cibc-mobile-b...   
1  https://app.appbot.co/apps/30102-cibc-mobile-b...   
2  https://app.appbot.co/apps/30102-cibc-mobile-b...   
4  https://app.appbot.co/apps/30102-cibc-mobile-b...   
5  https://app.appbot.co/apps/30102-cibc-mobile-b...   

                                              Topics  \
0                                                 []   
1                                                 []   
2                         ["Performance", "Pricing"]   
4  ["Frequency", "Security & Accounts", "Sign Up ...   
5                            ["Security & Accounts"]   

                    Custom Topics                                      Tags  
0                 ["Banking App"]                    ["Dissatisfied users"]  
1  ["Banking App", "Credit Card"]    ["Credit Score", "Dissatisfied users"]  
2                 ["Banking App"]                       ["Satisfied users"]  
4                              []  ["Dissatisfied users", "Password issue"]  
5                              []    ["Dissatisfied users", "Logout Issue"]

In [307]:
print('--2) Select the columns for analysis--\n')
Eng_reviews = Eng_reviews[['Review ID','App Store','Version','Rating','Subject','Body','Sentiment']]
Eng_reviews.head(10)


--2) Select the columns for analysis--



Review ID App Store  Version  Rating  \
0   3902085253       iOS  10.10.3       1   
1   3901178898       iOS  10.10.3       2   
2   3893450681       iOS  10.10.3       3   
4   3885434139       iOS  10.10.3       1   
5   3881274434       iOS  10.10.3       1   
6   3865788153       iOS  10.10.3       5   
7   3866472114       iOS  10.10.3       1   
8   3853019071       iOS  10.10.3       1   
9   3847909796       iOS  10.10.3       1   
10  3849972209       iOS  10.10.3       2   

                                 Subject  \
0                           Switch banks   
1               Credit Score Maintenance   
2                                   Good   
4                       Not good at all.   
5                                Log out   
6                       Cash back on app   
7                     would give 0 stars   
8   Ve r    y    S  l   o      w    now😴   
9                theres always problems!   
10                 Sign in with Face ID?   

                                                 Body Sentiment  
0         This app is as crappy as the bank that made  negative  
1   The credit score calculator has been down for ...  negative  
2   I stopped banking with CIBC due to another ban...  positive  
4   I have to change my password every time I try ...  negative  
5   Can’t log out. Or is very cumbersome to log ou...  negative  
6   I love the app. I never faced any issues so fa...  positive  
7   The absolute worst banking experience i’ve eve...  negative  
8   Very slow since last review, shield with the c...  negative  
9   it frequently wont allow me to transfer money ...  negative  
10  I’d like to have the option to sign into the a...  negative

In [308]:
print('--Merge Subject and Body into one column named Review--')
Eng_reviews['Review'] = Eng_reviews['Subject'].map(str) + ' ' + Eng_reviews["Body"].map(str)
Eng_reviews.head(10)


--Merge Subject and Body into one column named Review--


Review ID App Store  Version  Rating  \
0   3902085253       iOS  10.10.3       1   
1   3901178898       iOS  10.10.3       2   
2   3893450681       iOS  10.10.3       3   
4   3885434139       iOS  10.10.3       1   
5   3881274434       iOS  10.10.3       1   
6   3865788153       iOS  10.10.3       5   
7   3866472114       iOS  10.10.3       1   
8   3853019071       iOS  10.10.3       1   
9   3847909796       iOS  10.10.3       1   
10  3849972209       iOS  10.10.3       2   

                                 Subject  \
0                           Switch banks   
1               Credit Score Maintenance   
2                                   Good   
4                       Not good at all.   
5                                Log out   
6                       Cash back on app   
7                     would give 0 stars   
8   Ve r    y    S  l   o      w    now😴   
9                theres always problems!   
10                 Sign in with Face ID?   

                                                 Body Sentiment  \
0         This app is as crappy as the bank that made  negative   
1   The credit score calculator has been down for ...  negative   
2   I stopped banking with CIBC due to another ban...  positive   
4   I have to change my password every time I try ...  negative   
5   Can’t log out. Or is very cumbersome to log ou...  negative   
6   I love the app. I never faced any issues so fa...  positive   
7   The absolute worst banking experience i’ve eve...  negative   
8   Very slow since last review, shield with the c...  negative   
9   it frequently wont allow me to transfer money ...  negative   
10  I’d like to have the option to sign into the a...  negative   

                                               Review  
0   Switch banks This app is as crappy as the bank...  
1   Credit Score Maintenance The credit score calc...  
2   Good I stopped banking with CIBC due to anothe...  
4   Not good at all. I have to change my password ...  
5   Log out Can’t log out. Or is very cumbersome t...  
6   Cash back on app I love the app. I never faced...  
7   would give 0 stars The absolute worst banking ...  
8   Ve r    y    S  l   o      w    now😴 Very slow...  
9   theres always problems! it frequently wont all...  
10  Sign in with Face ID? I’d like to have the opt...

In [309]:
# Now let's check if Review has any empty value.
print('\n--Review has {} empty value(s).--\n'.format(Eng_reviews['Review'].isnull().sum()))


--Review has 0 empty value(s).--



In [320]:
print('--3) Remove punctuations, emojis--\n')

Eng_reviews['Review']=[s.encode('ascii', 'ignore').decode('ascii')\
                       for s in Eng_reviews['Review']]

print('--4) Normalize the words--\n')
Eng_reviews['Review']=Eng_reviews['Review'].str.lower()


print('--5) Remove punctuations--\n')
Eng_reviews['Review']=Eng_reviews['Review'].str.replace('[^A-Za-z ]+','')
#Eng_reviews['Review']=re.sub('[^A-Za-z ]+','',Eng_reviews['Review'])
Eng_reviews['Review'].head(10)


--3) Remove punctuations, emojis--

--4) Normalize the words--

--5) Remove punctuations--



0     switch banks this app is as crappy as the bank...
1     credit score maintenance the credit score calc...
2     good i stopped banking with cibc due to anothe...
4     not good at all i have to change my password e...
5     log out cant log out or is very cumbersome to ...
6     cash back on app i love the app i never faced ...
7     would give  stars the absolute worst banking e...
8     ve r    y    s  l   o      w    now very slow ...
9     theres always problems it frequently wont allo...
10    sign in with face id id like to have the optio...
Name: Review, dtype: object

In [323]:
print('-- Binary encode the sentiment--\n')
Eng_reviews['Sentiment_Encoded'] = [1 if x =='positive' else 0 for x in Eng_reviews.Sentiment]
Eng_reviews.head()

-- Binary encode the sentiment--



Review ID App Store  Version  Rating                   Subject  \
0  3902085253       iOS  10.10.3       1              Switch banks   
1  3901178898       iOS  10.10.3       2  Credit Score Maintenance   
2  3893450681       iOS  10.10.3       3                      Good   
4  3885434139       iOS  10.10.3       1          Not good at all.   
5  3881274434       iOS  10.10.3       1                   Log out   

                                                Body Sentiment  \
0        This app is as crappy as the bank that made  negative   
1  The credit score calculator has been down for ...  negative   
2  I stopped banking with CIBC due to another ban...  positive   
4  I have to change my password every time I try ...  negative   
5  Can’t log out. Or is very cumbersome to log ou...  negative   

                                              Review  Sentiment_encoded  \
0  switch banks this app is as crappy as the bank...                  0   
1  credit score maintenance the credit score calc...                  0   
2  good i stopped banking with cibc due to anothe...                  1   
4  not good at all i have to change my password e...                  0   
5  log out cant log out or is very cumbersome to ...                  0   

   Sentiment_Encoded  
0                  0  
1                  0  
2                  1  
4                  0  
5                  0

## Data Preprocessing

Now convert the Reviews into the X array and Sentiment into the y array accordingly.

In [350]:
from sklearn.model_selection import train_test_split
X, y =(Eng_reviews['Review'].values, Eng_reviews['Sentiment_Encoded'].values)
X_train_org, X_test_org, y_train_org, y_test_org = train_test_split(X, y, test_size = 0.25, random_state = 1)
X_test_org

array(['great app and great bank fantastic app and to everyone that complains about the e transfer or system issues its going to happen unfortunately they normally resolve it pretty fast',
       'love it love love love it does exactly what it says it will do  ',
       'doesnt work on rooted phones just because i want to root my phone and have control over what i have on my phone and what i dont have on my phone doesnt mean i shouldnt be able to run your app i will be deleting it forthwith i was even going to get one of your credit cards to use the app but i wont be now',
       ...,
       'app is not responsive version  crashes as soon as you sign init is frustrating cant use mobile version of site either because the tabs are not responsive what is going on cibc why the issue persists for so long pls fix',
       'nan last update to edeposits sucks big time thank gods i have a second device that hadnt updated yet',
       'nan this app would be much better if it had the capability o

In [352]:
from keras.preprocessing.text import text_to_word_sequence
from keras.preprocessing.sequence import pad_sequences

tk = Tokenizer(lower = True)
tk.fit_on_texts(X)
X_train_seq = tk.texts_to_sequences(X_train_org)
X_train_pad = pad_sequences(X_train_seq, maxlen=100, padding='post')

X_test_seq = tk.texts_to_sequences(X_test_org)
X_test_pad = pad_sequences(X_test_seq, maxlen=100, padding='post')


In [357]:
batch_size = 64
X_train1 = X_train_pad[batch_size:]
y_train1 = y_train_org[batch_size:]
X_valid = X_train_pad[:batch_size]
y_valid = y_train_org[:batch_size]

In [358]:
from keras.models import Sequential
from keras.layers import Embedding, LSTM, Dense, Dropout
vocabulary_size = len(tk.word_counts.keys())+1
max_words = 100
embedding_size = 32
model = Sequential()
model.add(Embedding(vocabulary_size, embedding_size, input_length=max_words))
model.add(LSTM(200))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [359]:
model.fit(X_train1, y_train1, validation_data=(X_valid, y_valid), batch_size=batch_size, epochs=10)

Train on 5486 samples, validate on 64 samples
Epoch 1/10
5486/5486 [==============================] - 35s 6ms/step - loss: 0.6613 - acc: 0.6358 - val_loss: 0.6755 - val_acc: 0.5938
Epoch 2/10
5486/5486 [==============================] - 33s 6ms/step - loss: 0.6563 - acc: 0.6376 - val_loss: 0.6825 - val_acc: 0.5938
Epoch 3/10
5486/5486 [==============================] - 33s 6ms/step - loss: 0.6526 - acc: 0.6376 - val_loss: 0.6782 - val_acc: 0.5938
Epoch 4/10
5486/5486 [==============================] - 33s 6ms/step - loss: 0.6498 - acc: 0.6376 - val_loss: 0.6852 - val_acc: 0.5938
Epoch 5/10
5486/5486 [==============================] - 33s 6ms/step - loss: 0.6424 - acc: 0.6376 - val_loss: 0.6880 - val_acc: 0.5938
Epoch 6/10
5486/5486 [==============================] - 33s 6ms/step - loss: 0.5499 - acc: 0.7346 - val_loss: 0.3609 - val_acc: 0.8750
Epoch 7/10
5486/5486 [==============================] - 34s 6ms/step - loss: 0.5626 - acc: 0.7386 - val_loss: 0.5788 - val_acc: 0.7031
Epoch 8/1

In [361]:
scores = model.evaluate(X_test_pad, y_test, verbose=0)
print('Test Accuracy:',scores[1])

Test Accuracy: 0.8416216213638718


In [362]:
y_predict = model.predict(X_test_pad)
y_predict

array([[0.8685466 ],
       [0.86854994],
       [0.16814384],
       ...,
       [0.16812989],
       [0.16812894],
       [0.16812901]], dtype=float32)

In [395]:
X_test_org=X_test_org.reshape(1850,1)
print(X_test_org.shape)
print(y_predict.shape)
final=np.concatenate([X_test_org,y_predict],axis=1)
df=pd.DataFrame(data=final, columns=['Review','Sentiment_Encoded'])


(1850, 1)
(1850, 1)


Review Sentiment
0     great app and great bank fantastic app and to ...  0.868547
1     love it love love love it does exactly what it...   0.86855
2     doesnt work on rooted phones just because i wa...  0.168144
3     touch id my touch id is not responding on the ...  0.168129
4     very convenient i find this app extremely conv...  0.168129
5     nan was five star the so called latest update ...  0.168129
6                                       nan easy to use  0.868549
7                                 wonderful awesomeness  0.868549
8                          crashes it crashesplease fix  0.168129
9     works sorta updated never clicking find us pre...  0.130488
10    nan for a shamelessly predatory greedy and une...  0.168129
11    wth lol i hate how you can only play if you ha...  0.168129
12                       functional does what it should  0.868548
13    nan i really enjoy using this bank very nice p...  0.168129
14    bill payments dont update balance remaining un...   0.16813
15    nan unfortunately mobile banking has stopped l...  0.168129
16    material design thank you finally finally fina...  0.868548
17    good but lacking crucial elements pros  really...  0.168129
18      touch id sign in is awesome new app is terrific  0.168129
19    fingerprint issues love the ability to use fin...  0.168129
20    allows easy viewing of obscene cibc fees app i...  0.168129
21    decent app but  i use this app but long for an...  0.168129
22    crash this thing is not operable after that up...   0.16813
23    money in the bank wow could i think of a lamer...   0.16813
24    mostly good one issue i like this app quick an...   0.16813
25                                 nan i think its cool  0.168129
26    please fix app was a great app till last updat...  0.168129
27                              nan what an amazing app  0.868547
28    nan this is not an app its just a webview dont...  0.168129
29    fast and easy to use use it everyday to check ...   0.86855
...                                                 ...       ...
1820                                           nan good  0.868549
1821                                      cibc great ap  0.868549
1822  app fail this new update is horrible app keeps...  0.168145
1823  white screen almost every time after i sign in...  0.168129
1824  edeposit finicky edeposit photo too sensitive ...  0.168129
1825                     equifax same here cant find it  0.168129
1826  must read this comment sketchy application thi...  0.127662
1827  nan didnt send a transfer  times had to pay a ...  0.168129
1828  need ipad version nice app but wish there is a...  0.168129
1829      the new app is amazing the new app is amazing  0.168129
1830                 sweet new look much more organized  0.868549
1831  almost perfect great app allows me to do almos...  0.868546
1832  nan app was fine until the update  same issue ...  0.168129
1833                                           nan good  0.868549
1834  asks for way too many permission why does this...  0.168129
1835                    nan works great very convenient   0.86855
1836  solid app solid and responsive app just a few ...  0.163389
1837  please fix the update crashes i actually liked...  0.168129
1838  wtf used to be a good app then i got the updat...  0.168129
1839  please make app work for pc financial this wou...  0.168129
1840                            cibc mobile banking app  0.868549
1841  basic app and still no iphone  or  plus suppor...  0.168129
1842  puts other banking apps to shame outstanding a...   0.86853
1843  whoever leads this team should be thanked grea...  0.168129
1844  app gone blanks app gone blank upon login plea...  0.168129
1845  might like it if i could get on it cant even l...  0.168129
1846    very good very easy to navigate simple and easy   0.86855
1847  app is not responsive version  crashes as soon...   0.16813
1848  nan last update to edeposits sucks big time th...  0.168129
1849  nan this app woul